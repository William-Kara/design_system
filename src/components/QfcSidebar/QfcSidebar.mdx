import { ArgTypes, Meta } from '@storybook/blocks';
import QfcSidebar from '@/components/QfcSidebar/QfcSidebar.vue';
import QfcButtonSidebar from '@/components/QfcButtonSidebar/QfcButtonSidebar.vue';

<Meta title="Modules/Sidebar" />

# Definition & Usage

## Description

The QfcSidebar component is a vertical menu placed on the left side of a webpage or application. It is used to navigate between different pages or sections of the application.

## Usage

To use QfcSidebar, style needs to be imported separately.

```sass
@import '@website/design-system-vuejs/dist/styles/components/QfcSidebar.scss';
```

```html
<template>
  <QfcSidebar v-model:isOpen="isSidebarOpen" :items="sidebarItems" />
</template>
```

## Props

The following props are available for the QfcSidebar component:

<ArgTypes of={QfcSidebar} />

Items prop take props as such:

<ArgTypes of={QfcButtonSidebar} />

- Please note that the `isOpen` prop is used to toggle the visibility of the sidebar menu.

- Note that the `screenId` prop can be set to `true` (any string) to enable the feature that pushes the main content
  screen when
  the sidebar is opened. The screenId prop is used to specify the id of the main content screen element. By default, this feature is **disabled**.
