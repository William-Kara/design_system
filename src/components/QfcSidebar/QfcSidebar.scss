.qfc-sidebar {
  height: 100%;
  width: 24rem;
  position: fixed;
  z-index: 1;
  top: 0;
  left: 0;
  background-color: $background;
  transition: $transition-time;
  box-shadow: 0 3px 6px #00000029;
  padding: 0 $q-generic-space-4;

  &__header {
    height: 6.25rem;
    margin-bottom: $q-generic-space-10;
    display: flex;
    justify-content: right;

    button {
      display: flex;
      align-items: center;
      background-color: transparent;
      border: none;
      color: $white;
      cursor: pointer;
    }
  }
  &--open {
    transform: translateX(-100%);
  }
}
