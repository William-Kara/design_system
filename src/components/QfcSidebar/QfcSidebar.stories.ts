import Sidebar from './QfcSidebar.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof Sidebar> = {
  component: Sidebar,
  title: 'Modules/Sidebar',
  argTypes: {
    items: {
      control: 'array',
      required: true,
      table: {
        type: {
          summary: 'Array<{ icon: string, label: string, to: string }>',
          detail: `
            {
              icon: string;
              label: string;
              to: string;
            }[]
          `,
        },
      },
    },
    isOpen: { control: 'boolean', required: true },
    screenId: {
      control: 'text',
      required: false,
      description: 'The ID of the main screen element.',
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    items: [
      {
        icon: 'space_dashboard',
        label: 'Dashboard',
        to: '/',
      },
      {
        icon: 'people',
        label: 'FAQ/Communaut√© Discord',
        to: '/',
      },
      {
        icon: 'settings',
        label: 'Settings',
        to: '/',
      },
    ],
    isOpen: true,
    screenId: 'storybook-root',
  },
};
