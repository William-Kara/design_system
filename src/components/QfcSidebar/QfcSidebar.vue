<template>
  <aside class="qfc-sidebar" :class="{ 'qfc-sidebar--open': !isOpen }">
    <div class="qfc-sidebar__header">
      <button @click="closeOnClick()">
        <i class="material-icons-outlined"> menu </i>
      </button>
    </div>
    <nav class="qfc-sidebar__nav">
      <QfcButtonSidebar
        v-for="({ label, to, icon }, index) in items"
        class="qfc-sidebar__item"
        :icon="icon"
        :key="index"
        :label="label"
        :to="to"
      />
    </nav>
  </aside>
</template>

<script lang="ts" setup>
import QfcButtonSidebar from '@/components/QfcButtonSidebar/QfcButtonSidebar.vue';
import { watch } from 'vue';

type ButtonSidebar = {
  icon: string;
  label: string;
  to: string;
};
interface Props {
  items: ButtonSidebar[];
  screenId: string;
  isOpen: boolean;
}
const props = withDefaults(defineProps<Props>(), {
  screenId: '',
  isOpen: false,
});

const emit = defineEmits(['update:isOpen']);
function closeOnClick() {
  emit('update:isOpen', false);
}

function togglePushOffScreen(screenId: string, transition: string = 'margin-left 0.5s') {
  const screen = document.getElementById(screenId);
  const sideBarWidth = document.querySelector('.qfc-sidebar')?.clientWidth.toString();

  if (screen && sideBarWidth) {
    screen.style.transition = transition;
    screen.style.marginLeft = props.isOpen ? sideBarWidth + 'px' : '0';
  }
}
if (props.screenId.length) {
  watch(
    () => props.isOpen,
    () => {
      togglePushOffScreen(props.screenId);
    }
  );
}
</script>

<style lang="scss" scoped src="./QfcSidebar.scss" />
