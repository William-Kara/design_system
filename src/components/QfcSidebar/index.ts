import QfcSidebar from './QfcSidebar.vue';

export { QfcSidebar };
export default QfcSidebar;
