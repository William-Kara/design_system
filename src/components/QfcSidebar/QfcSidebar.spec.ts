import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcSidebar from './QfcSidebar.vue';

let wrapper: any = null;

const propsValues = {
  items: [
    {
      icon: 'space_dashboard',
      label: 'Dashboard',
      to: '/dashboard',
    },
    {
      icon: 'people',
      label: 'FAQ/CommunautÃ© Discord',
      to: '/contacts',
    },
    {
      icon: 'settings',
      label: 'Settings',
      to: '/settings',
    },
  ],
  isOpen: false,
  screenId: 'storybook-root',
};

beforeEach(() => {
  wrapper = shallowMount(QfcSidebar, {
    propsData: propsValues,
  });
});

describe('QfcSidebar Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
