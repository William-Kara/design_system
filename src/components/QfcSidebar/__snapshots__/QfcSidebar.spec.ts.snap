// Vitest Snapshot v1

exports[`QfcSidebar Component > should render correctly 1`] = `
"<aside data-v-46033538=\\"\\" class=\\"qfc-sidebar qfc-sidebar--open\\">
  <div data-v-46033538=\\"\\" class=\\"qfc-sidebar__header\\"><button data-v-46033538=\\"\\"><i data-v-46033538=\\"\\" class=\\"material-icons-outlined\\"> menu </i></button></div>
  <nav data-v-46033538=\\"\\" class=\\"qfc-sidebar__nav\\">
    <qfc-button-sidebar-stub data-v-46033538=\\"\\" icon=\\"space_dashboard\\" label=\\"Dashboard\\" to=\\"/dashboard\\" class=\\"qfc-sidebar__item\\"></qfc-button-sidebar-stub>
    <qfc-button-sidebar-stub data-v-46033538=\\"\\" icon=\\"people\\" label=\\"FAQ/CommunautÃ© Discord\\" to=\\"/contacts\\" class=\\"qfc-sidebar__item\\"></qfc-button-sidebar-stub>
    <qfc-button-sidebar-stub data-v-46033538=\\"\\" icon=\\"settings\\" label=\\"Settings\\" to=\\"/settings\\" class=\\"qfc-sidebar__item\\"></qfc-button-sidebar-stub>
  </nav>
</aside>"
`;
